### PR TITLE
#158113796 Setup Database and Refactor code

### PR DESCRIPTION
**Pull request by asamisellah**
### What does this PR do?
Refactor the code to use blueprints and sets up connection of the app to the database 

### Description of Task to be completed?
- Install Postgres database(db)Create
- Connect db to api
- Create user table and requests table

** In refactor **
- Create a model file
- Create class User
- Create class Request
- Create user blueprint

### How should this be manually tested?
1. clone the [repo](https://github.com/asamisellah/Maintenance-Tracker)
2. open the folder in your text editor or IDE
3. `git checkout -b new-branch`
4. `git fetch ch-env-setup-157930685`

*Setup*
Make sure virtual environment is installed. Open terminal and run the following commands
1. `cd maintenance-tracker`
2.  `virtualenv venv` to create virtual environment
3.  `source venv/bin/activate` to run environment
4. `pip install -r requirements.txt`

*Database*
Prerequisites: Should have Postgres installed
Open terminal and run
``` $ sudo -u postgres psql
    $ CREATE DATABASE mtracker;
    $ \q   
    $ python schema.py
    $ python run.py
```
### Any background context you want to provide?
None

### What are the relevant pivotal tracker stories?
[#158141028](https://www.pivotaltracker.com/story/show/158141028/comments/190387102)

### CODE REVIEW

File: app/__init__.py [hunk 1]
```diff
@@ -1,110 +1,121 @@
...
+)
+from .model import User, UserRequest, get_all_requests
+from db_connect import TrackerDB
 
 app = Flask(__name__)
-
-requests = []
-users = []
-session = {}
+app.config['JWT_SECRET_KEY'] = 'super-secret'
```
Comment (daisymacharia) [inline app/__init__.py]: Is this your secret key? it should be secret (as the name suggests 😆 ) so put in in your env file and source it from there.

File: app/__init__.py [hunk 2]
```diff
@@ -1,110 +1,121 @@
...
 # POST a user
 @app.route('/api/v1/users', methods=['POST'])
 def create_user():
-    new_user = {
-        "user_id": str(uuid.uuid4()),
-        "username": request.json.get("username", "username"),
-        "email": request.json.get('email'),
-        "password": request.json.get("password")
-    }
+    new_user = model.User(
```
Comment (daisymacharia) [inline app/__init__.py]: why do you have to use the instance of User as `model.User`? I see you are importing it as User from the models, can't it just be accessed as User?

File: app/__init__.py [hunk 3]
```diff
@@ -1,110 +1,121 @@
...
+    )
     # Confirm user input has data
-    for key in new_user:
-        if new_user[key] == "":
-            return jsonify({"message": "All Fields Required"}), 400
+    if new_user.username is None:
+        return jsonify({"message": "All Fields Required"}), 400
+    elif new_user.email is None:
+        return jsonify({"message": "All Fields Required"}), 400
+    elif new_user.email is None:
```
Comment (daisymacharia) [inline app/__init__.py]: you are validating email twice, check `line 34`

File: app/__init__.py [hunk 4]
```diff
@@ -1,110 +1,121 @@
...
-        "user_id": str(uuid.uuid4()),
-        "username": request.json.get("username", "username"),
-        "email": request.json.get('email'),
-        "password": request.json.get("password")
-    }
+    new_user = model.User(
+        str(uuid.uuid4),
+        request.json.get("username"),
+        request.json.get("email"),
+        request.json.get("password")
```
Comment (daisymacharia) [inline app/__init__.py]: make sure you hash your passwords before saving them to the db you can do that after comparing your password and confirm password

File: app/__init__.py [hunk 5]
```diff
@@ -1,110 +1,121 @@
...
+    current_user = get_jwt_identity()
+    print(current_user)
+    request_data = UserRequest(
+        request.json.get("title"),
+        request.json.get('_type'),
+        request.json.get("description"),
+        request.json.get('category'),
+        request.json.get('area')
+    )
+    request_data.create_request()
```
Comment (daisymacharia) [inline app/__init__.py]: make sure you validate this data before you save it in your db

File: app/model.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+def get_user_requests(username):
+    db.cur.execute(
+        "SELECT * FROM requests WHERE username = ( % s)", (username,))
+    db.conn.commit()
+    requests = db.cur.fetchall()
+    return requests
+
+
+def get_user_request(username):
```
Comment (daisymacharia) [inline app/model.py]: this method is similar to `line 64` did you intend for this to be `get_request_by_id` ?

File: app/__init__.py (hunk 5 above)
Comment (asamisellah) [reply app/__init__.py]: I've validated the data for create request and update request, including signup user

File: app/__init__.py (hunk 2 above)
Comment (asamisellah) [reply app/__init__.py]: Aaah..you're right. I've made the change

File: app/__init__.py (hunk 1 above)
Comment (asamisellah) [reply app/__init__.py]: The key is secret now, for sure. Included it in an env file which is gitignored

File: app/__init__.py [hunk 7]
```diff
@@ -1,169 +1,197 @@
...
+    if len(user_request) != 0:
+        update_request(request_id,
+                       user_id,
+                       request.json.get("title").lower(),
+                       request.json.get("description").lower(),
+                       request.json.get("_type").lower(),
+                       request.json.get("category").lower(),
+                       request.json.get("area").lower()
+                       )
+        return jsonify({"Success": user_request}), 200
```
Comment (daisymacharia) [inline app/__init__.py:162]: remember the specifications are that a user should not be able to edit a request that has already been approved by the admin.

File: app/model.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+
+def get_user(username):
+    db.cur.execute("SELECT * FROM users WHERE username = (%s);", (username,))
+    db.conn.commit()
+    user = db.cur.fetchone()
+    return user
+
+
+class UserRequest():
```
Comment (daisymacharia) [inline app/model.py:52]: you are required to add a field for the request status that is only editable by the admin.

**Pull request merged**